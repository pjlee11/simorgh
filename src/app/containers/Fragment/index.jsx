import React from 'react';
import { string, node, arrayOf } from 'prop-types';

const italic = ({ children }) => <i>{children}</i>;
const bold = ({ children }) => <b>{children}</b>;

const attributeComponents = {
  italic,
  bold,
};

const fallbackAttributeComponent = ({ children }) => children;

const Fragment = ({ text, attributes }) =>
  /*
    Iterates through the attribute array and returns a component based on the attribute type (i.e. 'italic' or 'bold').
    These components are nested inside each other as children as the array is iterated through.
    The text string is passed in as the initial value, so it is the first child or the returned value if there are no attributes.
  */
  attributes.reduce((previousAttribute, attribute) => {
    const Attribute =
      attributeComponents[attribute] || fallbackAttributeComponent; // If attribute is unknown, will use a fallback component that just returns the passed children
    return <Attribute>{previousAttribute}</Attribute>;
  }, text);

Fragment.propTypes = {
  text: string.isRequired,
  attributes: arrayOf(string).isRequired,
};

bold.propTypes = {
  children: node.isRequired,
};

italic.propTypes = {
  children: node.isRequired,
};

export default Fragment;
