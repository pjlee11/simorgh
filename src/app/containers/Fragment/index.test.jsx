import React from 'react';
import { shouldMatchSnapshot } from '../../helpers/tests/testHelpers';
import Fragment from './index';

describe('Fragment', () => {
  describe('with no attributes', () => {
    shouldMatchSnapshot(
      'should render just text',
      <Fragment text="This is some text with no attributes" attributes={[]} />,
    );
  });

  describe('with bold attributes', () => {
    shouldMatchSnapshot(
      'should render text wrapped in a bold DOM element',
      <Fragment
        text="This is some text with bold attributes"
        attributes={['bold']}
      />,
    );
  });

  describe('with italic attributes', () => {
    shouldMatchSnapshot(
      'should render text wrapped in an italic DOM element',
      <Fragment
        text="This is some text with italic attributes"
        attributes={['italic']}
      />,
    );
  });

  describe('with bold and italic attributes', () => {
    shouldMatchSnapshot(
      'should render text wrapped in bold and italic DOM elements',
      <Fragment
        text="This is some text with bold and italic attributes"
        attributes={['bold', 'italic']}
      />,
    );
  });

  describe('with an unknown attribute', () => {
    shouldMatchSnapshot(
      'should ignore the attribute',
      <Fragment
        text="This is some text with a bold and unknown attribute"
        attributes={['bold', 'unknown']}
      />,
    );
  });
});
