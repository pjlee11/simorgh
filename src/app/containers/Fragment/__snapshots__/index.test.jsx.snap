// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Fragment with an unknown attribute should ignore the attribute 1`] = `
<b>
  This is some text with a bold and unknown attribute
</b>
`;

exports[`Fragment with bold and italic attributes should render text wrapped in bold and italic DOM elements 1`] = `
<i>
  <b>
    This is some text with bold and italic attributes
  </b>
</i>
`;

exports[`Fragment with bold attributes should render text wrapped in a bold DOM element 1`] = `
<b>
  This is some text with bold attributes
</b>
`;

exports[`Fragment with italic attributes should render text wrapped in an italic DOM element 1`] = `
<i>
  This is some text with italic attributes
</i>
`;

exports[`Fragment with no attributes should render just text 1`] = `"This is some text with no attributes"`;
